fe/fe-core/src/main/java/com/starrocks/qe/CoordinatorPreprocessor.java

```diff
@@ -623,25 +626,25 @@ private void computeFragmentHosts() throws Exception {
                 // hostSet contains target backends to whom fragment instances of the current PlanFragment will be
                 // delivered. when pipeline parallelization is adopted, the number of instances should be the size
                 // of hostSet, that it to say, each backend has exactly one fragment.
-                Set<TNetworkAddress> hostSet = Sets.newHashSet();
-
+                final Set<TNetworkAddress> hostSet = Sets.newHashSet();
+                final Set<TNetworkAddress> childHosts = Sets.newHashSet();
                 // don't use all nodes in shared_data running mode to avoid heavy exchange cost
                 if (connectContext.getSessionVariable().isPreferComputeNode() && hasComputeNode) {
-                    for (Map.Entry<Long, ComputeNode> entry : idToComputeNode.entrySet()) {
-                        ComputeNode computeNode = entry.getValue();
-                        if (!computeNode.isAlive() || SimpleScheduler.isInBlacklist(computeNode.getId())) {
-                            continue;
-                        }
-                        TNetworkAddress addr = new TNetworkAddress(computeNode.getHost(), computeNode.getBePort());
-                        hostSet.add(addr);
-                        this.recordUsedBackend(addr, computeNode.getId());
-                    }
-                    //make olapScan maxParallelism equals prefer compute node number
+                    Map<Long, ComputeNode> candidates = getAliveNodes(idToComputeNode);
+                    candidates.values().stream().forEach(e -> childHosts.add(e.getAddress()));
+                    // make olapScan maxParallelism equals prefer compute node number
+                    List<Pair<Long, TNetworkAddress>> chosenNodes = adaptiveChooseNodes(fragment, candidates, childHosts);
+                    chosenNodes.stream().forEach(e -> {
+                        hostSet.add(e.second);
+                        recordUsedBackend(e.second, e.first);
+                    });
                     maxParallelism = hostSet.size() * fragment.getParallelExecNum();
```

Comment:
introduce a variables like "plan_choose_node_mode = adaptive, locality or something in the future?"  this way clear the code and support more strageties in the future.